sdk/python/lib/pulumi/runtime/rpc.py

```diff
@@ -113,7 +116,14 @@
         return obj
 
     if inspect.isawaitable(value):
-        return await serialize_property(await value, deps, transform)
+        # Coroutines and Futures are both awaitable. Coroutines need to be scheduled.
+        # asyncio.ensure_future returns futures verbatim while converting coroutines into
+        # futures by arranging for the execution on the event loop.
+        #
+        # The returned future can then be awaited to yield a value, which we'll continue
+        # serializing.
+        future_return = await asyncio.ensure_future(value)
```

Comment:
This is subtle since it touches on how `asyncio` works in a pretty deep way:

`asyncio` has an ambient event loop (`asyncio.get_event_loop()` returns it) that is always running. `pulumi-language-python-exec` (our bootstrap script that loads a user's Pulumi program) initializes this event loop on startup and shuts it down when the program exits.

Unlike e.g. JavaScript, Python coroutines (`async def` functions) do nothing when called except return a coroutine object. In order to actually execute the function, it has to be scheduled on an event loop, which is what `asyncio.ensure_future` does here. `asyncio.ensure_future`, when given a coroutine as an argument, submits the coroutine for execution on the ambient event loop and returns a future that is completed whenever the coroutine completes.

So what we do here is call `asyncio.ensure_future` to schedule to coroutine, returning a future, and then we immediately `await` the future to produce a value (`future_return`).